src/Microsoft.AspNetCore.TestHost/ClientHandler.cs

```diff
@@ -187,27 +187,26 @@ internal void AbortRequest()
                 _responseStream.Complete();
             }
 
-            internal void CompleteResponse()
+            internal async Task CompleteResponseAsync()
             {
                 _pipelineFinished = true;
-                ReturnResponseMessage();
+                await ReturnResponseMessageAsync();
                 _responseStream.Complete();
-                _responseFeature.FireOnResponseCompleted();
+                await _responseFeature.FireOnResponseCompletedAsync();
             }
 
-            internal void ReturnResponseMessage()
+            internal async Task ReturnResponseMessageAsync()
             {
                 if (!_responseTcs.Task.IsCompleted)
                 {
-                    var response = GenerateResponse();
-                    // Dispatch, as TrySetResult will synchronously execute the waiters callback and block our Write.
-                    Task.Factory.StartNew(() => _responseTcs.TrySetResult(response));
```

Comment:
Please revert this, calling SetResult inline can cause threading issues and hangs